datafusion/physical-expr/src/expressions/not.rs

```diff
@@ -59,8 +63,25 @@ impl PhysicalExpr for NotExpr {
         self
     }
 
-    fn data_type(&self, _input_schema: &Schema) -> Result<DataType> {
-        Ok(DataType::Boolean)
+    fn data_type(&self, input_schema: &Schema) -> Result<DataType> {
+        // Ok(DataType::Boolean)
+        let data_type = self.arg.data_type(input_schema)?;
```

Comment:
That looks breaking imho... I would propose to have the own `BitwiseNotExpr` instead of making `NotExpr` universal

Reply:
As @alamb [suggestion](https://github.com/apache/arrow-datafusion/issues/5700#issuecomment-1481132344), I embed bitwise_not into NotExpr. But I agree that this way will break `NotExpr` original struct.  I think I should further discuss the implementation in the issue #5700

Reply:
ok, sounds good